content/extending-chat-widget/custom-identity-provider/index.mdx

```diff
@@ -0,0 +1,174 @@
+| `tokenType`    | Value: `Bearer`.                                                                 |
+| `creationDate` | Unix timestamp specyfing the time of creation of the token.                       |
+| `licenseId`    | LiveChat license ID.                                                             |
+
+## How to acquire it?
+
+In order to get the token you need to implement authorizing function making a call to a `/customer` endpoint from LiveChat Accounts API. To do so, you can follow [creating a new customer flow](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#creating-a-new-customer) or have a look at our [example](https://github.com/livechat/identity-provider-example/blob/master/lib/get-customer-token.ts) using cookies. Mind the fact that the [data](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#response-4) returned from this endpoint is not 1:1 with the token shape described above. You will have to add the `creationDate` and `licenseId` fields. Additionally, we expect `expiresIn` to be defined in milliseconds while the endpoint operates in seconds. Therefore you should multiply the provided value by `1000` before passing it further.
+
+
+# Examples
+
+Here are our suggested implementations of Customer Identity Provider. The `fetchLiveChatToken()` method used in examples is used to acquire the [Chat Widget token](#chat-widget-token).
+
+## Without caching
+
+If you don't need to cache the token, the fetching logic shall occur on every refresh of your application page. In such a case, the implementation can be very simple - we only have to ensure handling of `getToken` promise and simply resolve others.
+
+```js
+ window.__lc.custom_identity_provider = () => {
+    const fetchLiveChatToken = () => {
+        // fetch a token from LiveChat Accounts API here
+    }
+
+    let tokenPromise = null
+    const fetchToken = () => {
+        tokenPromise = fetchLiveChatToken()
+            .then(response => {
+                tokenPromise = null
+                return response
+            })
+    }
+
+    return {
+        getToken: () => tokenPromise || fetchToken(),
+        getFreshToken: () => Promise.resolve(),
+        hasToken: () => Promise.resolve(false),
+        invalidate: () => Promise.resolve()
+    }
+ }
+```
+
+## With caching
+
+If you care about the number of requests being made in your application, you can unleash the full potential of the Custom Identity Provider.
+
+```js
+window.__lc.custom_identity_provider = () => {
+    const fetchLiveChatToken = () => {
+        // fetch a token from LiveChat Accounts API here
+    }
+
+    const cacheKey = 'YOUR_CACHE_KEY'
+    let tokenPromise = null
+    let cachedToken = null
+
+    const isExpired = ({ creationDate, expiresIn }) => Date.now() >= creationDate + expiresIn
+
+    let retrievingToken = window.sessionStorage.getItem(cacheKey).then(token => {
```

Comment:
In what kind of an application the user of this solution could keep the token in the storage? It feels much safer to not do that ever and only keep the token in memory - after getting it from the backend.

Reply:
and why it would not be safe to store it? you can use it for example to handle page refreshes without the need for additional calls. You get expiration time in the response so you can check on your own if the token should still be valid and if no, get the fresh one. This is exactly how we are doing it and it is a pretty common practices.

Reply:
as @walaszczykm says its very common practice and well - we do so. This whole example is more advanced but provides mature approach to the problem. I've seen several authorization logic implementations and its not obvious for people how to handle concurrent promises (one from api and other from storage) so I went with showing the good example how can it be done. If someone is a pro, he/she will do it on its own as it will be understood, if someone is still learning - it will be really appreciated. Plus, it works for us and therefore guarantees it will work for clients so if someone messes with own approach we can just ¯\_(ツ)_/¯ and link this example